### DISCUSSION TITLE
Connect FattureInCloud and Excel

### DISCUSSION THREAD

**Discussion by raidatolokova** [fattureincloud/api#18] [Q&A] [answered]
Hi folks!
There is question, could someone help me to connect Excel and FattureInCloud? I am so far from programming and just received this task. I tried to create Application, then connected it, then received token. This token I inserted in excel in Data->Take data from Web, but I see so many different errors, so probbably I did everything wrong.
Relying on your help!

**Comment by valmoz** [comment]
Hi @raidatolokova 
first of all, welcome to the programming world! You'll indeed have some difficult or even frustrating moments, but it can also be really satisfying and give you a lot of possibilities. So stay strong, and you'll best it in no time!

Before starting, I have two suggestions.

1. try to describe your use case the best you can: what are you trying to do? what are the actions you're trying to perform with Fatture in Cloud? Are you creating this integration for internal use, or do you plan to resell it? we need more info to suggest to you how to proceed.
2. If you have [errors](https://developers.fattureincloud.it/docs/basics/errors), always send them to us. Usually, the errors contain details about what is going wrong, so knowing the error code and message can help us resolve your problem. Just be sure to not share sensitive data, and to replace them with fake data you made up as an example.
3. is it an [integration from scratch](https://developers.fattureincloud.it/docs/step-by-step), or are you [migrating a Version 1 integration to version 2](https://developers.fattureincloud.it/docs/api-v2-migration)?

Back to the general introduction. I'll try to keep it as simple as possible.

You can think of an API like if it is a barber shop: you have a set of actions you can ask to perform (shave, haircut, beard trim, ...) but you need to give instructions (how do you want your haircut, etc) and you have to pay for it.

Our menu can be found [here](https://developers.fattureincloud.it/api-reference).

As you said, the first thing to do is to authenticate. In REST APIs, the authorization demonstrates you can proceed with the request (like if you paid at the shop).

To get started, we think the easier way is to use the [Manual Auth](https://developers.fattureincloud.it/docs/authentication/manual-authentication). You'll be able to change auth method in the future if needed. 
Please, be sure to associate the right [scopes](https://developers.fattureincloud.it/docs/basics/scopes) to the token you create, otherwise, you'll get an error due to lack of permissions.

Once you obtain a token, you can start making requests. I suggest you to use VBA Macros for this (sic). Unfortunately, we don't provide VBA examples in our documentation, but you can just adjust the other programming language examples to your needs.
Our APIs are standard [REST APIs](https://developers.fattureincloud.it/docs/general-knowledge/rest-apis). This is excellent news for you because you can search on Google and you'll surely find a lot of documentation and code examples about REST.

To learn how to use REST APIs from VBA, you can search tutorials on Google, like the ones I found for you:

- https://youtu.be/pHWopg_XxMc
- https://youtu.be/KZeYKZJzQIk
- https://youtu.be/qE8r7S95oTg
- https://youtu.be/U-W_f7FsVz4
- https://youtu.be/2mR44X7PrRo
- https://youtu.be/a_RJzoj1cnM

Remember that the token you got from the authentication step must be inserted in a header as a bearer token when you perform REST requests:
https://www.appsloveworld.com/how-use-vba-to-get-data-from-api-with-bearer-token-to-excel

Once you selected the correct method (the action to perform) and authenticated the request, you must send instructions about how to do it. Every request has a set of parameters that you must set up to tell us how to perform the action.

If you don't know which action to perform, check out our [guides](https://developers.fattureincloud.it/docs/guides/).

I hope you'll find these examples useful. If you'll detail your use case more, we'll be able to help you to select the right methods you need.

Mauro

**Reply by raidatolokova** [reply]
Hi @valmoz! 
Wow, say that I am impressed is to say nothing! Thank you so much for the answer!

The idea is to connect FattureInCloud with Excel and export data about received and sent invoices, which later will be used in dashboards in PowerBI. It's all only for internal use.
What i was trying to do is through Excel->Data->Obtain data in Web to connect to FattureInCloud.

I will try now understand information that I received from you.
Again many thanks!

Raida

**Reply by valmoz** [reply]
You're welcome @raidatolokova 
then if it's only for internal use, feel free to use Manual Auth and ignore the OAuth2.0 part, you won't need it.

Let us know if you have further questions.
Mauro

**Reply by valmoz** [reply]
I forgot to mention: if you just need to export the received and issued document, then you will probably use the following methods:
- [List Issued Documents](https://developers.fattureincloud.it/api-reference#get-/c/-company_id-/issued_documents)
- [List Received Documents](https://developers.fattureincloud.it/api-reference#get-/c/-company_id-/received_documents)
You can try the methods directly from the [API Reference](https://developers.fattureincloud.it/api-reference#auth) (at the HTTP bearer section) before trying to implement them, so it will be easier to start. Just remember to insert the token at the top of the page to authenticate your requests.

You will most probably need to use one or more of the [Additional Capabilities](https://developers.fattureincloud.it/docs/basics/additional-capabilities/). Remember that you should retrieve the Company ID to proceed, as explained [at this link](https://developers.fattureincloud.it/docs/basics/company-scoped-methods).

**Reply by raidatolokova** [reply]
@valmoz 

What I could do based on my understanding:
1. Created application in FattureinCloud->Impostazioni->Sviluppatore
2. Connected application and received Access token
4. In API References at the HTTP bearer section entered Access token 
5. in the blocks: List Issued Documents and List Received Documents  I see some adequate responses
6. So from now on I should use VBA Macros?

Another question, previously I was trying to reach FattureInCloud from here. Could you please suggest something to me here? I receive following error messages: 

I really appreciate all your assistance and time!

Raida

**Reply by valmoz** [reply]
Hi @raidatolokova 
if you prefer to not write code at all, you can just try to use PowerQuery. Unfortunately, I'm not an expert on that tool, but I think you could try to check some external documentation like: 
- https://docs.microsoft.com/it-it/power-query/handlingauthentication
- https://www.codementor.io/@benniejayguevarra/importing-public-web-api-to-excel-no-coding-required-14bzfy4nev
- https://support.microsoft.com/it-it/office/importare-dati-dal-web-b13eed81-33fe-410d-9247-1747269c28e4
- https://support.microsoft.com/en-us/office/import-data-from-data-sources-power-query-be4330b3-5356-486c-a168-b68e9e616f5a
- https://raw-labs.com/blog/retrieving-json-data-from-a-rest-api-in-excel-with-power-query/

To be honest, I don't know if you'll be able to retrieve data from our APIs this way (because I don't know how you should set up the bearer token into it), but it's worth a try.

Please, notice that the URLs you're using are wrong. You should use:
- https://api-v2.fattureincloud.it/c/{company_id}/issued_documents
- https://api-v2.fattureincloud.it/c/{company_id}/received_documents
Remember to replace the company id in the previous URLs.

If you instead prefer to write some line of code, you have two main options:
- use VBA to access your data, as already explained (I think this is the best way)
- use some other programming language as an external script and save data to a CSV. Then import it into Excel

In the second case, you could use one of our supported languages (for example Python) and use our SDKs to make your job easier.

**Reply by valmoz** [reply]
BTW, NEVER share your key in a public discussion like this one.

**Reply by valmoz** [reply]
Hi @raidatolokova my colleague prepared a useful [quickstart to connect with our APIs using PowerQuery](https://developers.fattureincloud.it/docs/quickstarts/powerquery-quickstart). We hope it will be useful for you.

We plan to create one for VBA too in the future, but we don't have a deadline yet.

Let us know if you have further questions.

Thanks
Mauro

**Reply by raidatolokova** [reply]
Woooooooooooooooooooooooooooow!

Thank you so much! Such a wonderful news!
You cannot imagine how happy I am!

If this is possible can have someone's email, again I have question))) about authorization.

Sincerely yours and full of gratitude and admiration,
Raida

**Reply by valmoz** [reply]
Hi @raidatolokova,
if you have further questions you can ask us here on the community pages. If you need to share some sensible info that you can't share on a public discussion like this one, you can also contact our customer support on the Fatture in Cloud page.

[Here](https://developers.fattureincloud.it/docs/basics/authentication) you can find a high-level description of our auth mechanisms, and how to choose the best one for you.

Once you've chosen the one that best suits your needs, you can deep dive into reading one of the pages at [this link](https://developers.fattureincloud.it/docs/authentication/), where you'll find the details about each method.

You can also discover how to contact us following [this link](https://developers.fattureincloud.it/docs/support).

Thanks
Mauro

**Reply by valmoz** [reply]
Hi @raidatolokova do you still need some help? or can I close this discussion?

**Reply by raidatolokova** [reply]
Hi @valmoz,
I am super sorry to ask questions again...

By following an excel-query-quickstart, I received an answer that it is not possible to enter with assigned credentials. I tryed to use login and password of our Administrator in section "Basic" (in section "Windows" too). So I contacted Customer Support as was recommeded, so Customer support told me that I need first Bearer (before access token), where I can it?

![first step](https://user-images.githubusercontent.com/85202564/181832190-54c6b335-b115-4c5d-a683-5fbb71be71b7.jpg)
![second step](https://user-images.githubusercontent.com/85202564/181832214-36f79b58-613e-410c-a8bd-457d9e928697.jpg)

Excuse me please for all inconveniences.

**Reply by valmoz** [reply]
Hi @raidatolokova 
This time is much easier than you expect: you just have to prepend the prefix "Bearer " before the access token.
As shown here:
<img width="461" alt="image" src="https://user-images.githubusercontent.com/7254293/181835822-40cf378a-bd96-474b-9878-95844ebcfc10.png">

you don't have to replace it with something else: it is a constant string, so you can just add it as it is.

**Reply by raidatolokova** [reply]
@valmoz 
Oh, wow! 

this is what I received again(( I put one space between "Bearer" and access token
![error](https://user-images.githubusercontent.com/85202564/181837341-977bb035-ce6d-40a4-a686-1d12850ff2db.png)

**Comment by valmoz** [comment]
You must select "API Web" instead of anonima on the left

Inviato da Yahoo Mail su Android 
 
  Il ven, 29 lug, 2022 alle 22:20, ***@***.***> ha scritto:   


@valmoz
Oh, wow!

this is what I received again(( I put one space between "Bearer" and access token

**Reply by raidatolokova** [reply]
![errorAPIkey](https://user-images.githubusercontent.com/85202564/181839615-4aeeca4d-0e11-4e57-b1f3-30ac4f18178e.png)
Chiave is the same access token?

**Reply by valmoz** [reply]
Hi @raidatolokova 
unfortunately my Excel version doesn't have the web api import tool (probably because I'm using a Mac or for my kind of license), so I can't verify the current status of your integration on my own.
Maybe you can share a larger portion of your image? (you can hide the token if needed)

**Reply by raidatolokova** [reply]
Hi @valmoz!
I am back.
Here is the foto

**Reply by valmoz** [reply]
Hi @raidatolokova 
I'm really sorry, since I can't try on my mac I'm somewhat blind. I'll try with a colleague's windows PC as soon as possible.

It seems you were doing good in the previous attempts with the anonymous page. Could you give it another try as you were doing before (with the bearer thing) and let me know the result?

Please, remember to prepend the Bearer and check to have entirely copied the access token. To be totally sure not to have token issues, you can try to use the [API Reference](https://developers.fattureincloud.it/api-reference) to perform a REST call (in this case, the API reference tool inserts "Bearer" for you, so you must just paste the token as it is)

Thanks
Mauro

**Reply by valmoz** [reply]
Hi @raidatolokova 
I just asked my colleague to try the procedure, and it worked correctly.

Please, retry with the Anonymous part and the Bearer token. Be sure to write correctly the name of the authorization header and to copy the token in its entirety.

Thanks
Mauro

**Reply by raidatolokova** [reply]
Hi @valmoz 

I do not know what to say... SEEMS IT IS WORKING))

There is some information downloaded, we will read it carefully tomorrow and try to repeat procedure for [https://api-v2.fattureincloud.it/c/{company_id}/received_documents](https://api-v2.fattureincloud.it/c/%7Bcompany_id%7D/received_documents).

Thank you so much! You and your team are amazing!

**Reply by raidatolokova** [reply]
@valmoz 
I sincerely appreciate all your assistance and patience!

**Reply by valmoz** [reply]
Hi @raidatolokova 
thanks! great to hear that :)
Let us know if you have further questions!

**Reply by raidatolokova** [reply]
Hi @valmoz 

Hoped not to ask more questions, but...
in "Parti URL" will be like this?

https://api-v2.fattureincloud.it/
c/{company_id}/received_documents
?type=expenses

**Reply by valmoz** [reply]
Hi @raidatolokova 
as you can see [here](https://developers.fattureincloud.it/api-reference#get-/c/-company_id-/received_documents), the type is "expense".

**Comment by raidatolokova** [comment]
@valmoz 

Ehuu! Both are downloaded! Thank you! 
Now I can have a week of rest, hope you will have a vacation too)

Sincerely yours,
Raida

**Reply by valmoz** [reply]
@raidatolokova great! I'll make a summary of this discussion to help other users and I'll close this discussion.
Feel free to open another one if needed.

Mauro

**Comment by valmoz** [comment, marked as answer]
In order to connect to Fatture in Cloud APIs using Excel, you can use one of the following tools:

1. Power Query
2. VBA

We prepared a [quickstart using PowerQuery](https://developers.fattureincloud.it/docs/quickstarts/powerquery-quickstart), and a VBA quickstart will come soon.
